feat(gpu): KS32 support in GPU param sets, core-crypto GPU bindings, benches and tests

andrei-stoian-zama opened the pull request:
Change LWE keys to 32b and output LWEs to 32b, while input stays 64b.

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/zama-ai/tfhe-rs/2897)
<!-- Reviewable:end -->

agnesLeroy commented on backends/tfhe-cuda-backend/cuda/src/crypto/keyswitch.cuh:
```diff
@@ -165,16 +183,36 @@ __global__ void keyswitch_gemm_copy_negated_message_with_indices(
+    auto rounded_downscaled_body =
+        (KSTorus)(body_out >> input_to_output_scaling_factor);
+
+    body_out = -rounded_downscaled_body;
```
Same here @IceTDrinker, maybe someone else from the CPU team knows better about KS32 but I would appreciate to have an external pair of eyes to confirm this is the correct logic.

agnesLeroy commented on Makefile:
```diff
@@ -16,7 +16,7 @@ FAST_TESTS?=FALSE
 NIGHTLY_TESTS?=FALSE
 BENCH_OP_FLAVOR?=DEFAULT
-BENCH_TYPE?=latency
+BENCH_TYPE?=throughput
```
No need to change this I guess 🙂

IceTDrinker reviewed (commented):
The sign flip logic in the KS with type change in the cuda backend does not make sense at the moment to me @andrei-stoian-zama if you can comment on why you flip the body sign, I don't see the same thing in the CPU code

what we do for CPU as a first step is :

output_mask = 0
output_body = scale_down(round(input_body))

here you seem to flip the sign when setting the output body

andrei-stoian-zama commented on backends/tfhe-cuda-backend/cuda/src/crypto/keyswitch.cuh:
```diff
@@ -46,10 +46,23 @@ __device__ Torus *get_ith_block(Torus *ksk, int i, int level,
 
+template <typename T>
+__device__ T closest_repr(T input, uint32_t base_log, uint32_t level_count) {
+  T minus_2 = static_cast<T>(-2);
```
great idea! could you point out those rust tests that I should port to GPU ?

IceTDrinker replied:
can also take some of the doc test examples if that makes sense

agnesLeroy commented on backends/tfhe-cuda-backend/cuda/src/linearalgebra/multiplication.cuh:
```diff
@@ -203,11 +203,11 @@ __global__ void tgemm(uint M, uint N, uint K, const Torus *A, const Torus *B,
 // kernel from https://github.com/siboehm/SGEMM_CUDA
 // to any matrix dimension
-template <typename Torus, int BLOCK_SIZE, int THREADS>
+template <typename Torus, typename IndicesType, int BLOCK_SIZE, int THREADS>
```
Yes it's ok :+1:

andrei-stoian-zama commented:
> some gpu/silent shortint feature are in core in your latest patch
> 

Thanks. I think I now removed shortint references from core. Could you please check again?

With regards to the decomposers, I only implemented the "signedDecomposer" decomposition logic on GPU, as that's the one used in the KS32 implementation on CPU. Thus I only test SignedDecomposer. Would all decomposers pass the two tests I implemented ?

agnesLeroy commented on tfhe/src/shortint/atomic_pattern/compressed/ks32.rs:
```diff
@@ -73,6 +73,20 @@ impl CompressedKS32AtomicPatternServerKey {
         }
     }
 
+    pub fn into_raw_parts(
```
@andrei-stoian-zama it's still here

IceTDrinker commented:
*[`tfhe/src/high_level_api/keys/server.rs` line 491 at r5](https://reviewable.io/reviews/zama-ai/tfhe-rs/2897#-OgHuLUp9xnGJ3a1t5c7:-OgHuLUq1_lh8W6-4RPB:bjejthz) ([raw file](https://github.com/zama-ai/tfhe-rs/blob/9851f262caf50ea69dd65d8296a690b8cade140a/tfhe/src/high_level_api/keys/server.rs#L491)):*
> ```XML
>     pub fn gpu_indexes(&self) -> &[GpuIndex] {
>         match &self.key.key.key_switching_key {
>             CudaDynamicAtomicPatternKeySwitchingKey::KeySwitch32(ksk_32) => {
> ```

Still not the name I had in mind

CudaDynamicKeySwitchingKey

here it's more confusing I would think given DynamicAtomicPattern has another meaning for CPU

You can keep the original name, I was just merely pointing out the naming is maybe not the most accurate :)


<!-- Sent from Reviewable.io -->
